modelexpress_client/python/tests/test_refit_generator_client.py

```
@@ -779,6 +798,115 @@ def test_generator_retries_canonical_s3_under_one_lease(monkeypatch):
+
+def test_generator_rejects_replay_cycle_before_leasing(monkeypatch):
+    server, endpoint, service = _start_server()
+    service.version.CopyFrom(_canonical_version("version-b", "version-a"))
+    service.additional_versions = {
+        "version-a": _canonical_version("version-a", "version-b"),
+    }
+    adapter = _Adapter(service)
+    generator = _initialize(monkeypatch, endpoint, adapter, object_storage=True)
+
+    try:
+        with pytest.raises(RuntimeError, match="cycle detected.*version-b"):
```

Ruff (0.16.3)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)